### PR TITLE
eth: don't block if transaction broadcast loop is returned

### LINKED ISSUE #21167: personal.sendTransaction not broadcasting?
#### System information

Geth version: `Geth/v1.9.14-stable-6d74d1e5/linux-amd64/go1.14.2`
OS & Version: Linux

#### Expected behaviour
Transaction is broadcasted to the network and confirms

#### Actual behaviour
I get a tx hash in return, the tx is also added to the eth.pendingTransactions list, but never shows up in any ETH explorers and never confirms.

#### Steps to reproduce the behaviour
Try sending a transaction using personal.sendTransaction()

This is the output from eth.pendingTransactions if it helps:
```
}, {
    blockHash: null,
    blockNumber: null,
    from: "0x66ec231d3d50fd2c5a9a5242c76d91ca39a79c35",
    gas: 21000,
    gasPrice: 18010524411,
    hash: "0x9367ca24e37184f328a5b9edd67a776f98f2195803a6237d452de2e8f4fb2083",
    input: "0x",
    nonce: 5,
    r: "0x7f0e5d712072782ecb5ac39072044229d6283166a182749b0e50adce1224601f",
    s: "0x63c276600b15e7db1543fcb115c30cf94ee0948bbf7ead1200925602428ec54f",
    to: "0x4408f69a5a7ef19c5cf61309a046e735e2e38ea3",
    transactionIndex: null,
    v: "0x25",
    value: 1000000000000000
}]
```

Help would be really appreciated!

Update:
I've now tried this on both a light client and a fast sync. Both return a valid tx hash but it doesn't look like it's been broadcasted.

Update 2:
This issue occurs on both v1.9.9 and v1.9.14.

### PR DESCRIPTION
**Pull request by rjl493456442**
Fixes #21167

### CODE REVIEW

**Comment by rjl493456442** [comment]
```
goroutine 117 [select, 41 minutes]:
github.com/ethereum/go-ethereum/eth.(*peer).AsyncSendTransactions(0xc012aded00, 0xc167262180, 0x2, 0x2)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/eth/peer.go:362 +0xfb
github.com/ethereum/go-ethereum/eth.(*ProtocolManager).BroadcastTransactions(0xc0000fd960, 0xc011cd4000, 0xd, 0x10, 0x36b35588b338ae01)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/eth/handler.go:866 +0x4b5
github.com/ethereum/go-ethereum/eth.(*ProtocolManager).txBroadcastLoop(0xc0000fd960)
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/eth/handler.go:906 +0x163
created by github.com/ethereum/go-ethereum/eth.(*ProtocolManager).Start
	/home/travis/gopath/src/github.com/ethereum/go-ethereum/eth/handler.go:265 +0xc4
```

Check here for more information

**Comment by rjl493456442** [comment]
The root issue is whenever we failed to send transactions/transaction announces to a specific peer, the broadcast loop will return. https://github.com/ethereum/go-ethereum/blob/master/eth/peer.go#L185. Then all the following transaction broadcast requests will be blocked.

File: eth/peer.go [hunk 1]
```diff
@@ -368,7 +372,7 @@ func (p *peer) AsyncSendTransactions(hashes []common.Hash) {
 		for _, hash := range hashes {
 			p.knownTxs.Add(hash)
 		}
-	case <-p.term:
+	default:
```
Comment (karalabe) [inline eth/peer.go:375]: I'm not sure about the `default` change now that the failures drop the peer. Default branches should really only be used for notification channels when it's enough to add one item. Here we're doing data transfers, so a default will cause packets to get lost if the reader is not actively blocked on the read currently. I think we should revert these back to `p.term`.

File: eth/peer.go [hunk 2]
```diff
@@ -129,17 +129,19 @@ func newPeer(version int, p *p2p.Peer, rw p2p.MsgReadWriter, getPooledTx func(ha
 // broadcastBlocks is a write loop that multiplexes blocks and block accouncements
 // to the remote peer. The goal is to have an async writer that does not lock up
 // node internals and at the same time rate limits queued data.
-func (p *peer) broadcastBlocks() {
+func (p *peer) broadcastBlocks(removePeer func(string)) {
 	for {
 		select {
 		case prop := <-p.queuedBlocks:
 			if err := p.SendNewBlock(prop.block, prop.td); err != nil {
+				removePeer(p.id)
```
Comment (holiman) [inline eth/peer.go:137]: Is it safe to call removePeer from this thread?

Comment (holiman) [reply eth/peer.go:137]: Yeah, looks ok...

File: eth/peer.go [hunk 3]
```diff
@@ -129,17 +129,19 @@ func newPeer(version int, p *p2p.Peer, rw p2p.MsgReadWriter, getPooledTx func(ha
...
 				return
 			}
 			p.Log().Trace("Propagated block", "number", prop.block.Number(), "hash", prop.block.Hash(), "td", prop.td)
 
 		case block := <-p.queuedBlockAnns:
 			if err := p.SendNewBlockHashes([]common.Hash{block.Hash()}, []uint64{block.NumberU64()}); err != nil {
+				removePeer(p.id)
 				return
```
Comment (holiman) [inline eth/peer.go:145]: Is it really correct to `return` here? The `removePeer` should trigger `p.term` to close, which should exit the loop for us

Comment (karalabe) [reply eth/peer.go:145]: I'll try to rework it in the snap PR to not rely on a global channel that gets closed whenever rather on a per-broadcastre channel that remains in lockstep with the broadcaster methods.

**Pull request merged**